presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusRecordCursor.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    private Map<String, String> metricHeaderToMap(Map<String, String> mapToConvert)
+    {
+        return ImmutableMap.<String, String>builder().putAll(mapToConvert).build();
+    }
+
+    /**
+     * Encodes the given map into a Block.
+     *
+     * @param mapType Presto type of the map
+     * @param map     Map of key/value pairs to encode
+     * @return Presto Block
+     */
+    static Block getBlockFromMap(Type mapType, Map<?, ?> map)
+    {
+        // on functions like COUNT() the Type won't be a MapType
+        if (!(mapType instanceof MapType)) {
+            return null;
+        }
+        Type keyType = mapType.getTypeParameters().get(0);
+        Type valueType = mapType.getTypeParameters().get(1);
+
+        BlockBuilder mapBlockBuilder = mapType.createBlockBuilder(null, 1);
+        BlockBuilder builder = mapBlockBuilder.beginBlockEntry();
+
+        for (Map.Entry<?, ?> entry : map.entrySet()) {
+            writeObject(builder, keyType, entry.getKey());
+            writeObject(builder, valueType, entry.getValue());
+        }
+
+        mapBlockBuilder.closeEntry();
+        return (Block) mapType.getObject(mapBlockBuilder, 0);
+    }
+
+    /**
+     * Given the map type and Presto Block, decodes the Block into a map of values.
+     *
+     * @param type  Map type
+     * @param block Map block
+     * @return List of values
+     */
+    static Map<Object, Object> getMapFromBlock(Type type, Block block)
+    {
+        Map<Object, Object> map = new HashMap<>(block.getPositionCount() / 2);
+        Type keyType = Types.getKeyType(type);
+        Type valueType = Types.getValueType(type);
+        for (int i = 0; i < block.getPositionCount(); i += 2) {
+            map.put(readObject(keyType, block, i), readObject(valueType, block, i + 1));
+        }
+        return map;
+    }
+
+    /**
+     * Recursive helper function
+     *
+     * @param builder Block builder
+     * @param type    Presto type
+     * @param obj     Object to write to the block builder
+     */
+    static void writeObject(BlockBuilder builder, Type type, Object obj)
+    {
+        if (Types.isArrayType(type)) {
+            BlockBuilder arrayBldr = builder.beginBlockEntry();
+            Type elementType = Types.getElementType(type);
+            for (Object item : (List<?>) obj) {
+                writeObject(arrayBldr, elementType, item);
+            }
+            builder.closeEntry();
+        }
+        else if (Types.isMapType(type)) {
+            BlockBuilder mapBlockBuilder = builder.beginBlockEntry();
+            for (Map.Entry<?, ?> entry : ((Map<?, ?>) obj).entrySet()) {
+                writeObject(mapBlockBuilder, Types.getKeyType(type), entry.getKey());
+                writeObject(mapBlockBuilder, Types.getValueType(type), entry.getValue());
+            }
+            builder.closeEntry();
+        }
+        else {
+            TypeUtils.writeNativeValue(type, builder, obj);
```

Comment:
please verify that `obj` is of native type

Reply:
I see code similar to this in the Phoenix plugin, or is there a better way?
```suggestion
               if (BOOLEAN.equals(obj)
                    || TINYINT.equals(obj)
                    || SMALLINT.equals(obj)
                    || INTEGER.equals(obj)
                    || BIGINT.equals(obj)
                    || DOUBLE.equals(obj)) {
                TypeUtils.writeNativeValue(type, builder, obj);
            }
```

Reply:
well maybe not exactly that suggested code, since it causes test fail, but close to that maybe?

Reply:
> I see code similar to this in the Phoenix plugin

This looks good to me. Please also make sure that you throw an error when type does not fit any of case that you handle explicietly. In other words: `... else { throw ... }`

Reply:
Actually I'm having trouble with implementing this. What should be the matcher for VARCHAR for instance? I tried Presto's VARCHAR, but it's not matching on an object instance where I think it should...